File: pset2/readability/readability.c
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>

int get_clindex(string text);

int main(void)
{
    // Get text from user
    string text = get_string("Text: ");

    // Calculate Coleman-Liau index
    int index = get_clindex(text);

    // Print reading level
    if (index < 0)
    {
        printf("Before Grade 1\n");
    }
    else if (index < 16)
    {
        printf("Grade %i\n", index);
    }
    else
    {
        printf("Grade 16+\n");
    }
}

// Calculate Coleman-Liau index for given string
int get_clindex(string text)
{
    // Initalize index value and counters
    int letters = 0;
    int words = 1;
    int sentences = 0;

    // Loop through characters in text
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        char c = text[i];

        // If alphabetic add to letter count
        if (isalpha(c))
        {
            letters++;
        }

        // If space add to word count (exception for last word in sentence, therefore start words couner at 1)
        // Can assume sentences will not start or end with a space nor have multiple spaces in a row
        if (c == ' ')
        {
            words++;
        }

        // If period, exclamation point, or question mark add to sentence count
        // Can assume any seq. of chars ending with period, exclamaiton, or question mark to be a sentence
        if (c == '.' || c == '!' || c == '?')
        {
            sentences++;
        }
    }

    // printf("%i letters(s)\n", letters);
    // printf("%i words(s)\n", words);
    // printf("%i sentences(s)\n", sentences);

    // Calculate index and round to nearest whole number
    float L = letters * (100.0 / words);
    float S = sentences * (100.0 / words);
    int index = round(0.0588 * L - 0.296 * S - 15.8);

    return index;
}

File: week6/pset6/readability/readability.py
# Port of readability.c
from cs50 import get_string


def main():
    # Prompt user for text
    text = get_string("Text: ")
    while len(text) < 1:
        text = get_string("Text: ")

    # Calculate Coleman-Liau index
    index = calculate_cl(text)

    # Print grade
    if index < 1:
        print("Before Grade 1")
    elif index >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {round(index)}")


def calculate_cl(text):
    """Calculate Coleman-Liau index on text."""

    # Initialize counters
    letters = 0
    words = 1
    sentences = 0

    # Loop over text
    for char in text:
        if char.isalpha():
            letters += 1
        elif char in [".", "!", "?"]:
            sentences += 1
        elif char == " ":
            words += 1

    # Calculate index parameters
    L = 100 * letters / words
    S = 100 * sentences / words

    # Return index
    return 0.0588 * L - 0.296 * S - 15.8


if __name__ == "__main__":
    main()